### ISSUE TITLE
service --log-opt only works if you also specify --log-driver

### ISSUE THREAD

**Issue by Tomcat-Engineering** [docker/docs#8745]
### Problem description

If you run `docker service update myservice --log-opt mode=blocking` then it appears to run successfully but doesn't actually do anything.  In order to get it to set the log options, you also have to include the `--log-driver` flag in the same command.

This is mentioned in passing [here](https://github.com/moby/moby/issues/31515#issuecomment-321011661)

One day it would be nice if the command did actually honour the settings, but for now it would be helpful if the documentation mentioned the issue.

### Problem location

https://docs.docker.com/engine/reference/commandline/service_create/ and https://docs.docker.com/engine/reference/commandline/service_update/

### Suggestions for a fix

Add a comment in the documentation so that people don't spend ages wondering why their options aren't being applied.

**Comment by paigehargrave** [comment]
@thaJeztah Not sure if there's still an issue here (per https://github.com/moby/moby/issues/31515#issuecomment-321011661) - can you PTAL when you have a free moment (har har). Thanks!

**Comment by thaJeztah** [comment]
Thanks @Tomcat-Engineering (and thanks for the ping @paigehargrave)

Hm.. alright; so this looks to be a combination of something that's by design, and (what I think is) a bug in the `docker service update` logic; `docker service create` and `update` should follow slightly different logic (and I think they currently use the same).

- **expected**: setting `log-opt` for a service that does not have a logging-driver configured isn't supported
  - reason for this is https://github.com/moby/moby/issues/31515#issuecomment-321011661 that in a swarm setup, all instances of a service should use the same configuration (including logging-driver), so when specifying options, they must be accompanied with the driver  
- **t.b.d**.: creating a service with a log-opt set, but no logging driver _silently_ ignores the option
  - should it produce an error? a warning?
- **bug**: setting `log-opt` for a service that _has_ a logging driver configured (when created), but without specifying that driver during update is silently ignored.
  - when _updating_ a service, the CLI should take the existing logging-driver, and update the options


Taking a full example;

```bash
docker service create --log-driver=json-file --log-opt max-file=5 --name loggie nginx:alpine
docker service inspect --format='{{json .Spec.TaskTemplate.LogDriver}}' loggie
{"Name":"json-file","Options":{"max-file":"5"}}
```

Updating without logging-driver set doesn't take effect. IMO this should work, given that the service already had a logging-driver configured. It should still fail (or warn) if no logging-driver was configured yet on the service.

```bash
docker service update --log-opt max-file=15 loggie
docker service inspect --format='{{json .Spec.TaskTemplate.LogDriver}}' loggie
{"Name":"json-file","Options":{"max-file":"5"}}
```

Updating _with_ a logging driver specified does;

```bash
docker service update --log-driver=json-file --log-opt max-file=15 loggie
docker service inspect --format='{{json .Spec.TaskTemplate.LogDriver}}' loggie
{"Name":"json-file","Options":{"max-file":"15"}}
```

So, I agree that we should document this (or at least make sure that any example we have does use the right combination). This bug doesn't look critical (and there's a workaround), so likely won't find its way back in all older releases (e.g. likely won't be fixed in Docker Enterprise 17.06), so having this documented would be useful even if it might be fixed in a future update.


I'll copy the information here to the docker/cli issue tracker (where I _think_ the bug is), and will post a link to that issue here.

**Comment by thaJeztah** [comment]
opened https://github.com/docker/cli/issues/1942
